storage: work around segment index overflows

andrwng opened the pull request:
The segment index on offset deltas (difference between offset and segment's base offset) could overflow, resulting in very odd behavior, including returning ghost batches incorrectly which could result in lost batches during recovery.

To fix, this PR does two things:
- stops indexing offset deltas above `uint32_t::max`
- avoids using the index for offsets when we can tell the index spans more than `uint32_t::max` offsets

These changes feel pretty hacky, and our segment index still has a fundamental limitation for large segments. But this at least stops the bleeding.

<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [x] v25.1.x
- [x] v24.3.x
- [x] v24.2.x

## Release Notes

### Bug Fixes

* Fixes an issue with local segment indexes that could result in incorrect Raft recovery, compaction, or fetches in topic partitions with very large offsets (greater than 4294967295).
<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

secpanda commented:
![](https://res.cloudinary.com/snyk/image/upload/r-d/scm-platform/snyk-pull-requests/pr-banner-default.svg)

### :tada: **Snyk checks have passed. No issues have been found so far.**

:white_check_mark: **security/snyk** check is complete. No issues have been found. [(View Details)](https://app.snyk.io/org/devex-c3w/pr-checks/e943d9a4-b0e3-4dc3-810e-eb0f1d57775a?refs=fc1e8bfe727366cc41bc5e8feee9d64565312968&source=prChecksComment)

:white_check_mark: **license/snyk** check is complete. No issues have been found. [(View Details)](https://app.snyk.io/org/devex-c3w/pr-checks/e943d9a4-b0e3-4dc3-810e-eb0f1d57775a/license?refs=fc1e8bfe727366cc41bc5e8feee9d64565312968&source=prChecksComment)

secpanda commented:
![](https://res.cloudinary.com/snyk/image/upload/r-d/scm-platform/snyk-pull-requests/pr-banner-default.svg)

### :tada: **Snyk checks have passed. No issues have been found so far.**

:white_check_mark: **security/snyk** check is complete. No issues have been found. [(View Details)](https://app.snyk.io/org/devprod/pr-checks/0bf7f290-eae9-4061-ac60-21a02162f227?refs=fc1e8bfe727366cc41bc5e8feee9d64565312968&source=prChecksComment)

:white_check_mark: **license/snyk** check is complete. No issues have been found. [(View Details)](https://app.snyk.io/org/devprod/pr-checks/0bf7f290-eae9-4061-ac60-21a02162f227/license?refs=fc1e8bfe727366cc41bc5e8feee9d64565312968&source=prChecksComment)

andrwng committed:
storage: do not index past uint32_t::max

Once values begin coming in whose deltas are beyond uint32_t::max, the
index is basically unusable (queries have somewhat undefined behavior).

This updates the index to skip adding values if the offset delta would
overflow.

Prior to this change, one included test would fail with:

```
src/v/storage/tests/log_segment_reader_test.cc(376): error: in "test_ghost_read_with_index_overflow": check first.last_offset() == model::offset{100} has failed [2147483746 != 100]
src/v/storage/tests/log_segment_reader_test.cc(377): error: in "test_ghost_read_with_index_overflow": check first.header().type == model::record_batch_type::raft_data has failed [batch_type::ghost_batch != batch_type::raft_data]
```

This test failing means that a recovery from such an overflowed topic
could end up missing some batches.

Another included test would fail, this one relying on behavior addressed
by https://github.com/redpanda-data/redpanda/pull/25974 to cause
trouble:

```
src/v/storage/tests/log_segment_reader_test.cc(382): error: in "test_read_with_index_overflow_base": check first.base_offset() == model::offset{uint32_max} has failed [8589934591 != 4294967295]
src/v/storage/tests/log_segment_reader_test.cc(383): error: in "test_read_with_index_overflow_base": check first.last_offset() == model::offset{uint32_max} has failed [8589934591 != 4294967295]
```

This test failing means that a compaction could attempt to read a whole
segment but only get a part of a segment, and we could end up removing
batches incorrectly.

Just an interesting tidbit, these tests pass if using the batch cache.
This is relevant as we have been chasing issues with the
__consumer_offsets topic, which does not use the batch cache.

WillemKauf commented on src/v/storage/index_state.cc:
```diff
@@ -674,6 +677,14 @@ std::optional<index_state::entry> index_state::find_nearest(model::offset o) {
     if (o < base_offset || empty()) {
         return std::nullopt;
     }
+    long query_offset_delta = o() - base_offset();
```
Nit: prefer `int64_t` to long to keep with `model::offset` typedef

andrwng replied:
Done

WillemKauf commented on src/v/storage/index_state.cc:
```diff
@@ -674,6 +677,14 @@ std::optional<index_state::entry> index_state::find_nearest(model::offset o) {
     }
+    long query_offset_delta = o() - base_offset();
+    long seg_offset_delta = max_offset - base_offset;
+    constexpr long uint32_max = std::numeric_limits<uint32_t>::max();
```
nit: `static`

andrwng replied:
Done

WillemKauf reviewed (dismissed):
Whew. What a doozy

andrwng committed:
storage: don't use index with potential overflow

Indexes that have overflowed should not be used, because seeking within
is somewhat undefined when considering the potential of overflowed
uint32_t offset delta values.

This commit tweaks find_nearest() to return the first indexed entry if
this is detected. This allows Redpanda to safely proceed even when there
are segment indexes that have overflowed from previous versions of
Redpanda.

Without this change, the included test would fail with:

```
src/v/storage/tests/index_state_test.cc(360): Entering test case "index_overflow"
src/v/storage/tests/index_state_test.cc(377): error: in "index_overflow": check res->offset == model::offset{0} has failed [9 != 0]
src/v/storage/tests/index_state_test.cc(378): error: in "index_overflow": check res->filepos == 1 has failed [4 != 1]
src/v/storage/tests/index_state_test.cc(360): Leaving test case "index_overflow"; testing time: 77us
```

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/65276-->
<!--16a282ec-6002-41aa-94d1-1c8d878366c1-->
<!--abf2c0b7-bb85-476d-932b-ee27fdbca3fb-->
<!--246f972f-ab24-47ca-9c1e-8251c2d44a25-->
<!--8c215252-a0af-40be-8deb-eb5118c3db14-->
<!--de7a8193-69b9-4c1c-aa91-624862f822bd-->
<!--title-->
## Retry command for [Build#65276](https://buildkite.com/redpanda/redpanda/builds/65276)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```


/ci-repeat 1
tests/rptest/tests/data_migrations_api_test.py::DataMigrationsApiTest.test_conflicting_names
tests/rptest/tests/partition_reassignments_test.py::PartitionReassignmentsTest.test_add_partitions_with_inprogress_reassignments
tests/rptest/tests/scaling_up_test.py::ScalingUpTest.test_scaling_up_with_recovered_topic
tests/rptest/tests/datalake/datalake_e2e_test.py::DatalakeDelayedEnablementTest.test_enabling_iceberg_in_existing_cluster@{"catalog_type":"rest_jdbc","cloud_storage_type":1}
tests/rptest/tests/cloud_storage_timing_stress_test.py::CloudStorageTimingStressTest.test_cloud_storage@{"cleanup_policy":"compact,delete"}
tests/rptest/tests/simple_e2e_test.py::SimpleEndToEndTest.test_consumer_interruption
tests/rptest/tests/partition_balancer_test.py::PartitionBalancerTest.test_unavailable_nodes
```
<!--/body-->

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--ec6f0311-df3b-4dcc-9ff8-e6aa0b9415eb-->
<!--TEST RESULTS-->
<!--b94ca2d4-bd38-4bde-a620-8eaf96ac9db4-->
<!--TEST RESULTS-->
<!--3b356148-5d37-4d5f-99db-85ed5a605a30-->
<!--TEST RESULTS-->
<!--8b1aba7d-e4c5-4785-8cd3-0ab27426571e-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/65276">build#65276</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.cloud_storage_timing_stress_test.CloudStorageTimingStressTest.test_cloud_storage.cleanup_policy=compact.delete | ducktape | https://buildkite.com/redpanda/redpanda/builds/65276#019684fc-6af1-4674-a3c1-d5fa2ccefcfd | FLAKY | 19/21 |
| rptest.tests.data_migrations_api_test.DataMigrationsApiTest.test_conflicting_names | ducktape | https://buildkite.com/redpanda/redpanda/builds/65276#019684f8-7d09-48af-9d43-46f82536f8ff | FLAKY | 5/21 |
| rptest.tests.data_migrations_api_test.DataMigrationsApiTest.test_conflicting_names | ducktape | https://buildkite.com/redpanda/redpanda/builds/65276#019684fc-6af1-4674-a3c1-d5fa2ccefcfd | FLAKY | 5/21 |
| rptest.tests.datalake.datalake_e2e_test.DatalakeDelayedEnablementTest.test_enabling_iceberg_in_existing_cluster.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_JDBC | ducktape | https://buildkite.com/redpanda/redpanda/builds/65276#019684fc-6af1-4674-a3c1-d5fa2ccefcfd | FLAKY | 19/21 |
| rptest.tests.partition_reassignments_test.PartitionReassignmentsTest.test_add_partitions_with_inprogress_reassignments | ducktape | https://buildkite.com/redpanda/redpanda/builds/65276#019684fc-6aef-4820-9ee2-52d34a06e422 | FLAKY | 17/21 |
| rptest.tests.scaling_up_test.ScalingUpTest.test_scaling_up_with_recovered_topic | ducktape | https://buildkite.com/redpanda/redpanda/builds/65276#019684fc-6aef-4820-9ee2-52d34a06e422 | FLAKY | 9/21 |
| rptest.tests.simple_e2e_test.SimpleEndToEndTest.test_consumer_interruption | ducktape | https://buildkite.com/redpanda/redpanda/builds/65276#019684fc-6af1-4674-a3c1-d5fa2ccefcfd | FLAKY | 7/21 |
| rptest.tests.upgrade_test.UpgradeBackToBackTest.test_upgrade_with_all_workloads.single_upgrade=False | ducktape | https://buildkite.com/redpanda/redpanda/builds/65276#019684f8-7d08-42e4-8a0a-a5146febed3b | FLAKY | 16/21 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/65297">build#65297</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:-----------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.data_migrations_api_test.DataMigrationsApiTest.test_conflicting_names | ducktape | https://buildkite.com/redpanda/redpanda/builds/65297#019686f6-9e6e-423d-aec7-6d336cf8b575 | FLAKY | 4/21 |
| rptest.tests.simple_e2e_test.SimpleEndToEndTest.test_consumer_interruption | ducktape | https://buildkite.com/redpanda/redpanda/builds/65297#019686f6-9e71-4c7d-b78a-106963fa40c2 | FLAKY | 19/21 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/65323">build#65323</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:-----------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.data_migrations_api_test.DataMigrationsApiTest.test_conflicting_names | ducktape | https://buildkite.com/redpanda/redpanda/builds/65323#01968837-b14a-47d2-a472-60d46184f370 | FLAKY | 8/21 |
| rptest.tests.data_migrations_api_test.DataMigrationsApiTest.test_conflicting_names | ducktape | https://buildkite.com/redpanda/redpanda/builds/65323#0196883c-b316-49e9-952a-e5679a375bb3 | FLAKY | 7/21 |
| rptest.tests.partition_balancer_test.PartitionBalancerTest.test_recovery_mode_rebalance_finish | ducktape | https://buildkite.com/redpanda/redpanda/builds/65323#0196883c-b313-4800-a608-c1487204c841 | FLAKY | 20/21 |
| rptest.tests.partition_balancer_test.PartitionBalancerTest.test_unavailable_nodes | ducktape | https://buildkite.com/redpanda/redpanda/builds/65323#0196883c-b313-4800-a608-c1487204c841 | FLAKY | 16/21 |
| rptest.tests.partition_reassignments_test.PartitionReassignmentsTest.test_add_partitions_with_inprogress_reassignments | ducktape | https://buildkite.com/redpanda/redpanda/builds/65323#0196883c-b313-4800-a608-c1487204c841 | FLAKY | 20/21 |
| rptest.tests.simple_e2e_test.SimpleEndToEndTest.test_consumer_interruption | ducktape | https://buildkite.com/redpanda/redpanda/builds/65323#0196883c-b316-49e9-952a-e5679a375bb3 | FLAKY | 9/21 |
| rptest.tests.upgrade_test.UpgradeBackToBackTest.test_upgrade_with_all_workloads.single_upgrade=False | ducktape | https://buildkite.com/redpanda/redpanda/builds/65323#01968837-b147-4695-a358-3f113cdc6dd5 | FLAKY | 13/21 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/65386">build#65386</a></summary>


| test_id                                                                                                                                     | test_kind   | job_url                                                                                   | test_status   | passed   |
|:--------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.datalake.compaction_test.CompactionTest.test_compaction.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.NESSIE | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65386#01968b0d-edb6-4fba-96ab-c840210cb636 | FLAKY         | 7/21     |
| rptest.tests.partition_balancer_test.PartitionBalancerTest.test_unavailable_nodes                                                           | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65386#01968b0d-edb4-4354-8609-c5e3de214319 | FLAKY         | 9/21     |
| rptest.tests.upgrade_test.UpgradeBackToBackTest.test_upgrade_with_all_workloads.single_upgrade=False                                        | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65386#01968b0d-edb6-4fba-96ab-c840210cb636 | FLAKY         | 15/21    |
</details>
<!--/body-->

WillemKauf commented:
/ci-repeat 1
skip-redpanda-build
skip-units
tests/rptest/tests/data_migrations_api_test.py::DataMigrationsApiTest.test_conflicting_names
tests/rptest/tests/partition_reassignments_test.py::PartitionReassignmentsTest.test_add_partitions_with_inprogress_reassignments
tests/rptest/tests/scaling_up_test.py::ScalingUpTest.test_scaling_up_with_recovered_topic
tests/rptest/tests/datalake/datalake_e2e_test.py::DatalakeDelayedEnablementTest.test_enabling_iceberg_in_existing_cluster@{"catalog_type":"rest_jdbc","cloud_storage_type":1}
tests/rptest/tests/cloud_storage_timing_stress_test.py::CloudStorageTimingStressTest.test_cloud_storage@{"cleanup_policy":"compact,delete"}
tests/rptest/tests/simple_e2e_test.py::SimpleEndToEndTest.test_consumer_interruption
tests/rptest/tests/partition_balancer_test.py::PartitionBalancerTest.test_unavailable_nodes

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/65297-->
<!--a3beb791-b91b-4e44-be0f-be5041ba28d4-->
<!--title-->
## Retry command for [Build#65297](https://buildkite.com/redpanda/redpanda/builds/65297)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/data_migrations_api_test.py::DataMigrationsApiTest.test_conflicting_names
```
<!--/body-->

WillemKauf commented:
/ci-repeat 1
skip-redpanda-build
skip-units
tests/rptest/tests/data_migrations_api_test.py::DataMigrationsApiTest.test_conflicting_names

andrwng commented:
Thanks for the retries Willem!

dotnwat commented on src/v/storage/index_state.cc:
```diff
@@ -300,13 +300,16 @@ bool index_state::maybe_index(
-
-        retval = true;
+        auto offset_delta = batch_base_offset() - base_offset();
+        if (offset_delta <= std::numeric_limits<uint32_t>::max()) {
```
* is it possible for this to be false in the case where retval is true and would it matter? the initial case of the index the index being empty seems like an important edge case where we might always want the first entry?

* so skipping an entry is always safe right? it's only real effect is coarser granularity indexing?

andrwng replied:
>is it possible for this to be false in the case where retval is true and would it matter?

Ah hm, it is possible, but I don't think it matters: it just affects whether callers reset their respective accumulators, which affects the next time add_entry is evaluated, but presumably all entries at this point aren't going to be added. That said, it is an oversight to have this return true in the `|| retval` case. Will update.

>the initial case of the index the index being empty seems like an important edge case where we might always want the first entry?

Hm this is an interesting point. I think if the index stays empty, this is fine because queries into the index will return `std::nullopt` and callers should read from the start of the segment. And the index _should_ stay empty because we're relying on callers to call this with monotonically increasing records.

>so skipping an entry is always safe right? it's only real effect is coarser granularity indexing?

That's my read on it, yeah

bharathv commented on src/v/storage/index_state.cc:
```diff
@@ -299,16 +298,23 @@ bool index_state::maybe_index(
+        } else {
+            // We can't index anything beyond uint32 space. Presumably no
+            // further entries will be added because of this same condition.
+            return false;
```
curious if any instrumentation is useful around this (probably makes it easy to monitor all topics hitting such large offset spaces in one place).

edit: I'm thinking metrics.

andrwng replied:
We could probably add a metric for this, but I don't think one exists other than examining HWM for huge partitions.

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/65323-->
<!--95aa54b3-a369-4890-9a34-a69f8996041c-->
<!--2597c083-2ce0-4cab-86e3-7f4046493351-->
<!--f4b4c4d8-2412-444b-9394-3ebbb5eca0a2-->
<!--42a5ad17-995c-45f4-b69e-0e7deb3c6bf2-->
<!--fffed933-b67d-4133-b364-c47dd6dce11a-->
<!--e96a1a64-d9d2-41a8-9fe2-40a95f14f48d-->
<!--f70d36b7-a2c9-463f-851a-6bd1f4921ad8-->
<!--aec17ca7-5385-4bff-8045-4fcf757f92d6-->
<!--title-->
## Retry command for [Build#65323](https://buildkite.com/redpanda/redpanda/builds/65323)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```


/ci-repeat 1
tests/rptest/tests/data_migrations_api_test.py::DataMigrationsApiTest.test_conflicting_names
tests/rptest/tests/upgrade_test.py::UpgradeBackToBackTest.test_upgrade_with_all_workloads@{"single_upgrade":false}
tests/rptest/tests/simple_e2e_test.py::SimpleEndToEndTest.test_consumer_interruption
tests/rptest/tests/partition_reassignments_test.py::PartitionReassignmentsTest.test_add_partitions_with_inprogress_reassignments
tests/rptest/tests/partition_balancer_test.py::PartitionBalancerTest.test_recovery_mode_rebalance_finish
tests/rptest/tests/partition_balancer_test.py::PartitionBalancerTest.test_unavailable_nodes
tests/rptest/tests/cloud_storage_timing_stress_test.py::CloudStorageTimingStressTest.test_cloud_storage_with_partition_moves@{"cleanup_policy":"compact,delete"}
```
<!--/body-->

secpanda commented:
![](https://res.cloudinary.com/snyk/image/upload/r-d/scm-platform/snyk-pull-requests/pr-banner-default.svg)

### :tada: **Snyk checks have passed. No issues have been found so far.**

:white_check_mark: **security/snyk** check is complete. No issues have been found. [(View Details)](https://app.snyk.io/org/cloud-1t5/pr-checks/3fcc58f5-ef2a-4f03-888a-8ed0a6d75ffb?refs=fc1e8bfe727366cc41bc5e8feee9d64565312968&source=prChecksComment)

:white_check_mark: **license/snyk** check is complete. No issues have been found. [(View Details)](https://app.snyk.io/org/cloud-1t5/pr-checks/3fcc58f5-ef2a-4f03-888a-8ed0a6d75ffb/license?refs=fc1e8bfe727366cc41bc5e8feee9d64565312968&source=prChecksComment)

andrwng commented:
Latest [push](https://github.com/redpanda-data/redpanda/compare/7446cbd046630b25b2abe7d78918bdfbb7d63e63..87706cdc8b0e3055cbb400046fcf01be5b2404b0) adds a test case that would previously fail to read a full segment, implying this really is a cause of trouble in compaction.

andrwng commented:
Latest push: [rebase](https://github.com/redpanda-data/redpanda/compare/87706cdc8b0e3055cbb400046fcf01be5b2404b0..66d27b208a0af4cf58e980118edfe0fabe384afa)

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/65386-->
<!--95a9fadf-63f3-405c-8e2d-5623c40d504b-->
<!--title-->
## Retry command for [Build#65386](https://buildkite.com/redpanda/redpanda/builds/65386)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/datalake/compaction_test.py::CompactionTest.test_compaction@{"catalog_type":"nessie","cloud_storage_type":1}
```
<!--/body-->

andrwng commented:
/ci-repeat 1
skip-redpanda-build
skip-units
tests/rptest/tests/datalake/compaction_test.py::CompactionTest.test_compaction@{"catalog_type":"nessie","cloud_storage_type":1}

vbotbuildovich commented:
/backport v25.1.x

vbotbuildovich commented:
/backport v24.3.x

vbotbuildovich commented:
/backport v24.2.x

vbotbuildovich commented:
Failed to create a backport PR to v24.2.x branch. I tried:

```
git remote add upstream https://github.com/redpanda-data/redpanda.git
git fetch --all
git checkout -b backport-pr-25979-v24.2.x-177 remotes/upstream/v24.2.x
git cherry-pick -x 0ed8d64254 fc1e8bfe72
```

[Workflow run logs](https://github.com/redpanda-data/redpanda/actions/runs/14784586534).
